optuna/storages/_rdb/storage.py

```diff
@@ -1132,6 +1147,59 @@ def get_all_versions(self) -> List[str]:
 
         return self._version_manager.get_all_versions()
 
+    def record_heartbeat(self, trial_id: int) -> None:
+        with _create_scoped_session(self.scoped_session, True) as session:
+            heartbeat = models.TrialHeartbeatModel.where_trial_id(trial_id, session)
+            if heartbeat is None:
+                heartbeat = models.TrialHeartbeatModel(trial_id=trial_id)
+                session.add(heartbeat)
+            else:
+                heartbeat.heartbeat = session.execute(func.now()).scalar()
+
+    def fail_stale_trials(self) -> List[int]:
+        stale_trial_ids = self._get_stale_trial_ids()
+
+        for trial_id in stale_trial_ids:
+            self.set_trial_state(trial_id, TrialState.FAIL)
+
+        return stale_trial_ids
+
+    def _get_stale_trial_ids(self) -> List[int]:
+        assert self.heartbeat_interval is not None
+        if self.grace_period is None:
+            grace_period = 2 * self.heartbeat_interval
+        else:
+            grace_period = self.grace_period
+        stale_trial_ids = []
+
+        with _create_scoped_session(self.scoped_session, True) as session:
+            current_heartbeat = session.execute(func.now()).scalar()
+            current_heartbeat = current_heartbeat.replace(tzinfo=None)
```

Comment:
I think we should leave a comment for this line, especially since it's specific to PostgreSQL. For a more curious readers, here's a thread from sqlalchemy that might give some background into why this happens. I'm not aware of any better workaround. https://github.com/sqlalchemy/sqlalchemy/issues/275